### ISSUE TITLE
Running average metric

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#264]
Following the discussion in #256 on running average values in the TQDM progress bar, IMO it would be better to split running average computation and display logic.

Here is an implementation on how to compute running average value from an instance of Metric:
```python
from ignite.metrics import Metric


class RunningAverage(Metric):
    """Compute running average of a metric
    """
    
    def __init__(self, metric, alpha=0.98):
        assert isinstance(metric, Metric), "Argument metric should be an instance of Metric"
        assert 0.0 < alpha <= 1.0, "Argument alpha should be a float between 0.0 and 1.0"
        self.metric = metric
        self.alpha = alpha
        super(RunningAverage, self).__init__()
        
    def reset(self):
        self._value = None

    def iteration_completed(self, engine):        
        self.metric.iteration_completed(engine)

    def compute(self):
        if self._value is None:
            self._value = self.metric.compute()
        else:
            self._value = self._value * self.alpha + (1.0 - self.alpha) * self.metric.compute()
        return self._value
    
    def attach(self, engine, name):
        # restart average every epoch
        engine.add_event_handler(Events.EPOCH_STARTED, self.started)
        # reset metric at each iteration        
        engine.add_event_handler(Events.ITERATION_COMPLETED, self.metric.started)
        # compute metric
        engine.add_event_handler(Events.ITERATION_COMPLETED, self.iteration_completed)
        # apply running average
        engine.add_event_handler(Events.ITERATION_COMPLETED, self.completed, name)
```

Usage of the proposed API is following:
```python
import numpy as np
import torch

from ignite.engine import Engine, Events, State
from ignite.metrics import Loss, CategoricalAccuracy


n_iters = 100
batch_size = 10
n_classes = 10
data = list(range(n_iters))
y_true_batch_values = iter(np.random.randint(0, n_classes, size=(n_iters, batch_size)))
y_pred_batch_values = iter(np.random.rand(n_iters, batch_size, n_classes))
loss_values = iter(range(n_iters))


def update_fn(engine, batch):
    loss_value = next(loss_values)
    y_true_batch = next(y_true_batch_values)
    y_pred_batch = next(y_pred_batch_values)
    return loss_value, torch.from_numpy(y_pred_batch), torch.from_numpy(y_true_batch)


trainer = Engine(update_fn)

alpha = 0.98
acc_metric = RunningAverage(CategoricalAccuracy(output_transform=lambda x: [x[1], x[2]]), alpha=alpha)
acc_metric.attach(trainer, 'running_avg_accuracy')

loss_metric = RunningAverage(Loss(torch.nn.CrossEntropyLoss(), output_transform=lambda x: [x[1], x[2]]), alpha=alpha)
loss_metric.attach(trainer, 'running_avg_loss')

@trainer.on(Events.ITERATION_COMPLETED)
def log_running_avg_metrics(engine):
    print("running avg accuracy:", engine.state.metrics['running_avg_accuracy'])
    print("running avg loss:", engine.state.metrics['running_avg_loss'])    
    
```

There is however still missing the part of the running average value based on `state.output`.
Maybe we could find a solution together by improving the previous code...

What do you guys think about ?
Thanks 

cc @miguelvr

**Comment by miguelvr** [comment]
not being able to compute it over engine.state.output can be an annoyance. But I guess we can solve it with an `output_transform` function, right?

**Comment by vfdev-5** [comment]
@miguelvr could you please give an example of API with `output_transform` on engine.state.output ?
I'm not sure to understand your idea

**Comment by miguelvr** [comment]
we could make `RunningAverage` a `handler` instead of a `Metric`. I would be something that processes `metrics` rather than being one on itself. 

Plus, if you make it with an `output_tranform` argument, you are able to access both `engine.state.metrics` and `engine.state.output`. 

What do you think?

**Comment by vfdev-5** [comment]
I see what you mean. I thought about such implementation, maybe it worth to try to implement

**Comment by vfdev-5** [comment]
Here is naive implementation of `RunningAverage` with joint `Metric` and output:
```python
from ignite.metrics import Metric


class RunningAverage(object):
    """Compute running average of a metric or output
    """
    
    def __init__(self, src, alpha=0.98, output_transform=None):
        assert isinstance(src, Metric) or src == 'output', "Argument src should be a Metric or string 'output'"
        assert 0.0 < alpha <= 1.0, "Argument alpha should be a float between 0.0 and 1.0"

        self.src = src
        if isinstance(src, Metric):
            assert output_transform is None, "Argument output_transform should be None if src is 'output'"
            self._get_src_value = self._get_metric_value
            self.iteration_completed = self._metric_iteration_completed
        else:
            self._get_src_value = self._get_output_value
            self.iteration_completed = self._output_iteration_completed

        self.alpha = alpha
        self._output_transform = output_transform
        super(RunningAverage, self).__init__()
    
    def reset(self, engine):
        self._value = None
        
    def _metric_iteration_completed(self, engine):
        self.src.started(engine)
        self.src.iteration_completed(engine)

    def _output_iteration_completed(self, engine):
        self.src = self._output_transform(engine.state.output)
        
    def completed(self, engine, name):
        if self._value is None:
            self._value = self._get_src_value()
        else:
            self._value = self._value * self.alpha + (1.0 - self.alpha) * self._get_src_value()    
        engine.state.metrics[name] = self._value

    def attach(self, engine, name):
        # restart average every epoch
        engine.add_event_handler(Events.EPOCH_STARTED, self.reset)
        # compute metric
        engine.add_event_handler(Events.ITERATION_COMPLETED, self.iteration_completed)
        # apply running average
        engine.add_event_handler(Events.ITERATION_COMPLETED, self.completed, name)

    def _get_metric_value(self):
        return self.src.compute()
        
    def _get_output_value(self):
        return self.src
```
with API usage is similar to the previous one
```python
acc_metric = RunningAverage(CategoricalAccuracy(output_transform=lambda x: [x[1], x[2]]), alpha=alpha)
acc_metric.attach(trainer, 'running_avg_accuracy')

loss_metric = RunningAverage(Loss(torch.nn.CrossEntropyLoss(), output_transform=lambda x: [x[1], x[2]]), alpha=alpha)
loss_metric.attach(trainer, 'running_avg_loss')

avg_output = RunningAverage('output', output_transform=lambda x: x[0], alpha=alpha)
avg_output.attach(trainer, 'running_avg_output')

@trainer.on(Events.ITERATION_COMPLETED)
def log_running_avg_metrics(engine):
    print("running avg accuracy:", engine.state.metrics['running_avg_accuracy'])
    print("running avg loss:", engine.state.metrics['running_avg_loss'])
    print("running avg output:", engine.state.metrics['running_avg_output'])

```

I think we can leave `RunningAverage` as being a `Metric` and simplify a little bit the code...

**Comment by vfdev-5** [comment]
> I think we can leave RunningAverage as being a Metric and simplify a little bit the code...

Actually no real gain of such inheritance. Will leave as a handler.

**Comment by alykhantejani** [comment]
I think it makes sense for this to inherit from Metric. Besides simplifying the code a little bit, the running average (of a metric, or output value) is a metric we are computing, no?

**Comment by vfdev-5** [comment]
@alykhantejani I see your point, for a coherent implementation it would better to inherit from Metric as we write in all cases in engine.state.metric and maybe `RunningAverage` is better to be in metrics module. But in the code I have no impression of any gain with the inheritance

**Comment by miguelvr** [comment]
how would you compute it over `engine.state.output` then?

**Comment by vfdev-5** [comment]
@miguelvr it can be something like this:
```python
avg_output = RunningAverage(output_transform=lambda x: x[0], alpha=alpha)
# or RunningAverage(src=None, output_transform=lambda x: x[0], alpha=alpha)
avg_output.attach(trainer, 'running_avg_output')
```
the result is in `engine.state.metrics['running_avg_output']`

**Comment by miguelvr** [comment]
nice! I like it
